pkg/workload/workload.go

```diff
@@ -188,15 +188,14 @@ func FindConditionIndex(status *kueue.WorkloadStatus, conditionType string) int
 	return -1
 }
 
-// UpdateStatus updates the condition of a workload.
+// UpdateStatus updates the condition of a workload with ssa,
+// Each condition has it's own field manager with the same name
```

Comment:
The field manager should be per controller, not per condition.
However, this means that we need to send in the patch all the conditions that are owned by the controller.

The kube client has a tool for obtaining all the fields of a kubernetes API, but I don't know if there is an equivalent for CRDs.

https://kubernetes.io/blog/2021/08/06/server-side-apply-ga/#using-server-side-apply-in-a-controller

Reply:
Indeed the field management can be done at a controller level, the problem is , as you mentioned , the need to extract the other field of interest every time a patch operation is done. 

From an execution point of view, having a dedicated manager for every option gives us the opportunity to modify the options concurrently. With the current approach the one think we are not "guarded" against is the concurrent update "PodsReady" and    "Finished" by the job controller.

What the core types use:
1. Code generated by `applyconfiguration-gen` which is not trivial to integrate.
2. `k8s.io/apimachinery/pkg/util/managedfields.ExtractInto` used to extract the values of a fieldManager from an existing option.

`managedfields.ExtractInto` could by used without any code generators (by using  `typed.DeducedParseableType` ) , however the conditions slice for our scenario is treated as `atomic` hence no concurrency in updates.

Reply:
> From an execution point of view, having a dedicated manager for every option gives us the opportunity to modify the options concurrently.

But we don't need such control. Each controller process a particular object just once at a time. So there shouldn't be race conditions among workers of the same controller.

>  however the conditions slice for our scenario is treated as atomic hence no concurrency in updates.

This doesn't do the job?

https://github.com/kubernetes-sigs/kueue/blob/cf76bfe02ae58e5005403821689592c6b967ce5a/apis/kueue/v1alpha2/workload_types.go#L106

Having a manager per field deviates from the recommendations for how to use SSA. I'm checking with some people whether there are any risks that we should be aware of. One I can think of is that more names implies more chances of collision if users write their own controllers to manipulate Workloads.

Reply:
> This doesn't do the job? 
It does in the API server, however  typed.DeducedParseableType can only use reflection and only has access to the `json` tags .

Reply:
What is the output of `.metadata.managedFields` after the first patch?

Reply:
> the conditions slice for our scenario is treated as atomic hence no concurrency in updates

atomic vs list is a schema setting and the server and client (extract flow) should agree about it or we have a major problem. If this is a CRD you can change the type by putting the tag in the schema. If it is currently atomic your plan to use a different field manager for each condition type isn't going to work :)

You should get the openapi schema from the server, not use a deduced schema.

cc @apelisse

Reply:
In general, I think I would advise for that specific pattern here for conditions. Conditions are often independent from each others (and so they can't be all applied at once), there's usually O(3-5) which is perfectly fine. Having separate manager to handle each condition is fine, maybe add the controller name along with the name of the fieldmanager, otherwise you might run into trouble if you continue to use that pattern (multiple controllers editing the same condition with the same fieldmanager).

Reply:
There is no problem with the type definition, the behavior I was referring to is  just the effect of using `typed.DeducedParseableType`

> You should get the openapi schema from the server, not use a deduced schema.

@lavalamp  Can you give me some pointers about how could I do that?

@alculquicondor  here is  the managedFields section
```
[
  {
    "apiVersion": "kueue.x-k8s.io/v1beta1",
    "fieldsType": "FieldsV1",
    "fieldsV1": {
      "f:status": { "f:admission": { "f:clusterQueue": {}, "f:podSetFlavors": { "k:{\"name\":\"main\"}": { ".": {}, "f:flavors": { "f:cpu": {}, "f:memory": {} }, "f:name": {} } } } }
    },
    "manager": "kueue-admission",
    "operation": "Apply",
    "subresource": "status",
    "time": "2023-03-09T09:10:47Z"
  },
  {
    "apiVersion": "kueue.x-k8s.io/v1beta1",
    "fieldsType": "FieldsV1",
    "fieldsV1": {
      "f:status": { "f:conditions": { "k:{\"type\":\"Admitted\"}": { ".": {}, "f:lastTransitionTime": {}, "f:message": {}, "f:reason": {}, "f:status": {}, "f:type": {} } } }
    },
    "manager": "kueue-admission-Admitted",
    "operation": "Apply",
    "subresource": "status",
    "time": "2023-03-09T09:10:47Z"
  },
  {
    "apiVersion": "kueue.x-k8s.io/v1beta1",
    "fieldsType": "FieldsV1",
    "fieldsV1": {
      "f:status": { "f:conditions": { "k:{\"type\":\"PodsReady\"}": { ".": {}, "f:lastTransitionTime": {}, "f:message": {}, "f:reason": {}, "f:status": {}, "f:type": {} } } }
    },
    "manager": "kueue-job-controller-PodsReady",
    "operation": "Apply",
    "subresource": "status",
    "time": "2023-03-09T09:10:50Z"
  },
  {
    "apiVersion": "kueue.x-k8s.io/v1beta1",
    "fieldsType": "FieldsV1",
    "fieldsV1": {
      "f:status": { "f:conditions": { "k:{\"type\":\"Finished\"}": { ".": {}, "f:lastTransitionTime": {}, "f:message": {}, "f:reason": {}, "f:status": {}, "f:type": {} } } }
    },
    "manager": "kueue-job-controller-Finished",
    "operation": "Apply",
    "subresource": "status",
    "time": "2023-03-09T09:11:22Z"
  },
  {
    "apiVersion": "kueue.x-k8s.io/v1beta1",
    "fieldsType": "FieldsV1",
    "fieldsV1": {
      "f:metadata": {"...."},
      "f:spec": {"...."}
    },
    "manager": "kueue",
    "operation": "Update",
    "time": "2023-03-09T09:10:47Z"
  }
]

```

Reply:
Thanks. It looks like ownership is properly partitioned, matching to the mapType.

@apelisse is mapType something that is supported by `managedfields.ExtractInto` and `typed.DeducedParseableType`?

Reply:
I'm confused, do you still want to use `ExtractInto`? Because no it wouldn't work.

Reply:
If it doesn't work, I'm ok having a manager per condition for now.

Reply:
I think it'll be significantly simpler